plugins/pentaho-googledrive-vfs/core/src/main/java/org/pentaho/googledrive/vfs/GoogleDriveFileObject.java

```diff
@@ -0,0 +1,236 @@
+  private HttpTransport HTTP_TRANSPORT;
+  private String APPLICATION_NAME = "";
+  private Drive driveService;
+  private FileType mimeType;
+  private String id;
+
+  public enum MIME_TYPES {
+    FILE( "application/vnd.google-apps.file", FileType.FILE ), FOLDER( "application/vnd.google-apps.folder",
+        FileType.FOLDER );
+    private final String mimeType;
+    private final FileType fileType;
+
+    private MIME_TYPES( String mimeType, FileType fileType ) {
+      this.mimeType = mimeType;
+      this.fileType = fileType;
+    }
+
+    public static FileType get( String type ) {
+      FileType fileType = null;
+      if ( FOLDER.mimeType.equals( type ) ) {
+        fileType = FOLDER.fileType;
+      } else {
+        fileType = FILE.fileType;
+      }
+      return fileType;
+    }
+  }
+
+  private final java.io.File DATA_STORE_DIR = new java.io.File( resolveCredentialsPath() );
+
+  protected GoogleDriveFileObject( final AbstractFileName fileName, final GoogleDriveFileSystem fileSystem )
+      throws FileSystemException {
+    super( fileName, fileSystem );
+    try {
+      HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
+      DATA_STORE_FACTORY = new CustomDataStoreFactory( DATA_STORE_DIR );
+      driveService = getDriveService();
+      resolveFileMetadata();
+    } catch ( Exception e ) {
+      throw new FileSystemException( e );
+    }
+  }
+
+  protected String[] doListChildren() throws Exception {
+    String[] children = null;
+    if ( isFolder() ) {
+      id = id == null ? "root" : id;
+      String fileQuery = "'" + id + "' in parents and trashed=false";
+      FileList files = driveService.files().list().setQ( fileQuery ).execute();
+      List<String> fileNames = new ArrayList<String>();
+      for ( File file : files.getFiles() ) {
+        fileNames.add( file.getName() );
+      }
+      children = fileNames.toArray( new String[0] );
+    }
+    return children;
+  }
+
+  protected void doCreateFolder() throws Exception {
+    if ( !getName().getBaseName().isEmpty() ) {
+      File folder = new File();
+      folder.setName( getName().getBaseName() );
+      folder.setMimeType( MIME_TYPES.FOLDER.mimeType );
+      folder = driveService.files().create( folder ).execute();
+      if ( folder != null ) {
+        id = folder.getId();
+        mimeType = MIME_TYPES.get( folder.getMimeType() );
+      }
+    }
+  }
+
+  protected void doDelete() throws Exception {
+    driveService.files().delete( id ).execute();
+    id = null;
+    mimeType = null;
+  }
+
+  protected long doGetContentSize() throws Exception {
+    return -1;
+  }
+
+  protected FileType doGetType() throws Exception {
+    return mimeType;
+  }
+
+  protected InputStream doGetInputStream() throws Exception {
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    driveService.files().get( id ).executeMediaAndDownloadTo( out );
+    ByteArrayInputStream in = new ByteArrayInputStream( out.toByteArray() );
```

Comment:
This seems exceedingly dangerous. You're copying the complete contents of the file into memory which is sure to lead to an OOME which is difficult to recover from. Definitely shouldn't copy the stream to a BAOS.